IDE-13400 add events section (#139)
intergiro/developer-portal: docs/3d/common/events.md

Content-Type: application/json
Authorization: Bearer <access_token>
```

In this case we are choosing ascending order to retrieve the oldest events first, and we specify the last seen
event id.

```{1,17,25}
HTTP 200 OK

{
  "data": [
    {
      "created_at": "2022-08-16T12:45:33.000Z"
      "id": "b1ded866-79cc-44d2-aa61-69c26a2f580a",
      "name": "ExampleEvent",
      "payload": {
        "individual_id": "bd4af0ff-0a72-4e09-b247-c6d8c2391496",
        "example_field": "Example value"
      }
    },
    ...
    {
      "created_at": "2022-08-17T13:51:10.000Z"
      "id": "42943401-f3cb-4c33-a2ed-ffe87055d227",
      "name": "ExampleEvent",
      "payload": {
        "individual_id": "a04bb96a-0203-48a1-8dcf-00a13aaafec9",
        "example_field": "Example value"
      }
    }
  ],
  "has_more": true
}
```

We save the `id` of the last event in the list. Since the `has_more` field has value `true`,
your system should repeat this process until it is `false`.
At that point, the next polling interval should begin with last saved cursor and begin the process again.

### Webhooks

Webhooks serve as a method of immediate notification about an event being created.

> They do not contain any data from event payload except for the id, event name and timestamp.

Reviewer: ```suggestion
They do not contain any data from event payload except for the `id`, `name` and `created_at` fields.
```

::: warning
Webhooks should be used as an addition to [polling for events](#polling), never a replacement.
:::

#### Registering new url

> In order to register a new webhook url, make a request to the following endpoint:

Reviewer: ```suggestion
In order to register a new webhook URL, make a request to the following endpoint:
```

```{1}
POST /v3/webhooks

Content-Type: application/json
Authorization: Bearer <access_token>

{
  "url": "https://example.com/webhook-url"
}
```

> You can register at most one webhook URL. Make sure the URL you are registering uses HTTPS.

Reviewer: Perhaps we should explain why using HTTPS is a requirement

Reviewer: @knowakowski-intergiro No. Please correct me if its incorrect, but don't we require that the webhook url is a HTTPS scheme url, or do we also allow HTTP scheme url's?

If we only allow HTTPS schema url's, it may be worth noting why this is the case, i.e. security reasons

I think that explaining what HTTPS does and its advantages over HTTP is out of scope, and not something we should explain :)

Author: Yes, we require HTTPS scheme URLs only.
So the reason for that would be for the benefits HTTPS provides over HTTP (like you mentioned the added security), there is nothing specific to our system there.

Reviewer: Yes, you may be right. I rest the case and resolve the comment :)

#### Listing webhooks

You can retrieve the webhooks you've registered to verify their URL.

```{1}
GET /v3/webhooks

Content-Type: application/json
Authorization: Bearer <access_token>
```

Response:

```{1}
HTTP 200 OK

[
  {
    "created_at": "2022-07-01T13:17:19.000Z",
    "id": "a6495db9-8d9d-417d-9918-c5b4a122e712",
    "url": "https://example.com/webhook-url"
  }
]
```

#### Removing registered URL

> In order to remove an existing webhook url, you need to have the webhook ID which is retrievable by

Reviewer: ```suggestion
In order to remove an existing webhook URL, you need to have the webhook ID which is retrievable by
```

Content-Type: application/json
Authorization: Bearer <access_token>
```

#### Receiving webhooks

When a new event is created, a webhook is dispatched to the specified URL.

Example webhook payload:

```json
{
  "id": "b1ded866-79cc-44d2-aa61-69c26a2f580a",
  "name": "ExampleEvent",
  "created_at": "2022-08-16T12:45:33.000Z"
}
```

To retrieve the event associated with the webhook, use the `id` from the webhook's body.

```{1}
GET /v3/events/b1ded866-79cc-44d2-aa61-69c26a2f580a

Content-Type: application/json
Authorization: Bearer <access_token>
```

In the response you will receive a `payload` field which contains data relevant to the event.

```{1,6-9}
HTTP 200 OK

{
  "id": "b1ded866-79cc-44d2-aa61-69c26a2f580a",
  "name": "ExampleEvent",
  "payload": {
  	"individual_id": "bd4af0ff-0a72-4e09-b247-c6d8c2391496",
  	"example_field": "Example value"
  },
  "created_at": "2022-08-16T12:45:33.000Z"
}
```

#### Responding to webhooks

Whenever your system receives a request to the webhook URL,
it should acknowledge it by responding with a `2XX` HTTP code.

If your system fails to respond with a `2XX` HTTP code or if there are any network disruptions,

> the request will be retried at most 5 times with an exponential backoff base of 2 seconds.

Reviewer: Should we explain a simple example of the exponential backoff? What do you think?

Author: Good idea, done.
